astroquery/utils/commons.py

```diff
@@ -453,3 +453,22 @@ def parse_votable(content):
     """
     tables = votable.parse(six.BytesIO(content), pedantic=False)
     return tables
+
+
+def _open_shelve(shelffn, withclosing=False):
+    """
+    Opens a shelf file.  If `withclosing` is  True, it will be opened with closing,
+    allowing use like:
+
+        with _open_shelve('somefile',True) as s:
+            ...
+    """
+    import shelve
+    import contextlib
+
+    shelf = shelve.open(shelffn, protocol=2)
+
+    if withclosing:
+        return contextlib.closing(shelf)
```

Comment:
At a glance, I don't see why this would behave differently than `__exit__()` in https://github.com/python/cpython/blob/master/Lib/shelve.py . Maybe it is something subtle in the implementation details. Academic discussions for another day...